common/types.go

```diff
@@ -168,3 +169,71 @@ type LogConfig struct {
 	FileCount   uint64
 	TimeTracker bool
 }
+
+type Block struct {
+	Id         string
+	StartIndex int64
+	EndIndex   int64
+	Size       int64
+}
+
+// list that holds blocks containing ids and corresponding offsets
+type BlockOffsetList struct {
+	BlockOffsetList []*Block //blockId to offset mapping
+	SmallFile       bool     // does it consist of blocks
+	Cached          bool     // is it cached?
+}
+
+func (bol BlockOffsetList) binarySearch(offset int64) (bool, int) {
+	lowerBound := 0
+	higherBound := len(bol.BlockOffsetList) - 1
+	for lowerBound <= higherBound {
+		middleIndex := (lowerBound + higherBound) / 2
+		// we found the starting block that changes are being applied to
+		if bol.BlockOffsetList[middleIndex].EndIndex > offset && bol.BlockOffsetList[middleIndex].StartIndex <= offset {
+			return true, middleIndex
+			// if the end index is smaller or equal then we need to increase our lower bound
+		} else if bol.BlockOffsetList[middleIndex].EndIndex <= offset {
+			lowerBound = middleIndex + 1
+			// if the start index is larger than the offset we need to decrease our upper bound
+		} else if bol.BlockOffsetList[middleIndex].StartIndex > offset {
+			higherBound = middleIndex - 1
+		}
+	}
+	return false, 0
+}
+
+func (bol BlockOffsetList) FindBlocksToModify(offset, length int64) (*BlockOffsetList, int64, bool) {
+	// size of mod block list
+	size := int64(0)
+	currentBlockOffset := offset
+	modBlockList := BlockOffsetList{}
+	found, index := bol.binarySearch(offset)
+	if !found {
+		return &modBlockList, 0, true
+	}
+	// after the binary search just iterate to find the remaining blocks
+	for _, blk := range bol.BlockOffsetList[index:] {
+		if blk.StartIndex > offset+length {
+			break
+		}
+		if currentBlockOffset >= blk.StartIndex && currentBlockOffset < blk.EndIndex && currentBlockOffset <= offset+length {
```

Comment:
Instead of generating a list of blocks to be modified, as blocks are always in ascending order of offset we can just return the first block from where the modification starts and then the write method can just start from that block and keep modifying until length of write operation and doing the same it can create new block if required. This will simplify the logic a bit I feel.